### PR TITLE
fix(seldon operator): infinite service updating if extra default fileds applied by k8s

### PR DESCRIPTION
**Pull request by caozhuozi**
**What this PR does / why we need it**:
This PR tries to fix issue https://github.com/SeldonIO/seldon-core/issues/4689. The issue and the cause are detailed described there.

**Which issue(s) this PR fixes**:

Fixes (https://github.com/SeldonIO/seldon-core/issues/4689)

**Special notes for your reviewer**:

1. As I metioned in issue  (https://github.com/SeldonIO/seldon-core/issues/4689), the root cause of the infinite updating services in every reconcile is due to the extra fileds applied by k8s. To solve this, I enhanced the check statement by just comparing those required fields and ignoring the default fields  applied by k8s when comapring the desired svc and the found one. Specifically,

   * Since several fields need to be compared/updated, I organized the check/update logic into  separate functions.
   * As the code has already ignored the extra fields set by k8s,  I also removed the [line 1261](https://github.com/SeldonIO/seldon-core/blob/8004de80ec447d5337d78ef630328584b3177829/operator/controllers/seldondeployment_controller.go#L1261) `svc.Spec.ClusterIP = found.Spec.ClusterIP`.
2. I think there is also a potential bug at [line 1282](https://github.com/SeldonIO/seldon-core/blob/8004de80ec447d5337d78ef630328584b3177829/operator/controllers/seldondeployment_controller.go#L1282). The deep copy should be performed **after** the update to `found`. Otherwise, the comparing logic at [line 1293](https://github.com/SeldonIO/seldon-core/blob/8004de80ec447d5337d78ef630328584b3177829/operator/controllers/seldondeployment_controller.go#L1293) is meaningless. Why we re campare the old found svc(stored in `desiredSvc`) with the updated one(`found` is subsequently updated **twice** at [line 1284-1285](https://github.com/SeldonIO/seldon-core/blob/8004de80ec447d5337d78ef630328584b3177829/operator/controllers/seldondeployment_controller.go#L1284-L1285) and [line 1287](https://github.com/SeldonIO/seldon-core/blob/8004de80ec447d5337d78ef630328584b3177829/operator/controllers/seldondeployment_controller.go#L1287) sperately) given that we already know they are not the same? I think this behavior also conflicts with what the code comment at [line 1292](https://github.com/SeldonIO/seldon-core/blob/8004de80ec447d5337d78ef630328584b3177829/operator/controllers/seldondeployment_controller.go#L1293) intends to do: 
    > // Check if what came back from server modulo the defaults applied by k8s is the same or not
    
    In my opinion, "what came back from server" refers to the second update to `found` at [line 1287](https://github.com/SeldonIO/seldon-core/blob/8004de80ec447d5337d78ef630328584b3177829/operator/controllers/seldondeployment_controller.go#L1287), that is, 
    ```go
    err = r.Update(context.TODO(), found)
    ```
    If I understand correctly, the comment originally wants to compare the desired Service  just before sending update request to the API server and the result back from the server.
    By moving the deep copy statement just after the first update to `found` and before the second update, I think the comapring statement now matches the what the above comment wants to do as well as the variable name itself: **desired**Svc( I was very confused about why the deep copy of the old found one is called `desiredSvc`).
    Please corrrect me if I'm wrong. Thanks.

### CODE REVIEW

**Comment by seldondev** [comment]
Hi @caozhuozi. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) or [todo](https://github.com/orgs/todo/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/todo/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

File: operator/controllers/seldondeployment_controller.go [hunk 1]
```diff
@@ -866,6 +866,39 @@ func createContainerService(deploy *appsv1.Deployment,
...
+//   - Spec.Type
+//   - Spec.Selector
+//   - Spec.SessionAffinity
+func compareServices(desiredSvc, found *corev1.Service) bool {
+	return equality.Semantic.DeepEqual(desiredSvc.Annotations, found.Annotations) &&
+		equality.Semantic.DeepEqual(desiredSvc.Labels, found.Labels) &&
+		equality.Semantic.DeepEqual(desiredSvc.Spec.Ports, found.Spec.Ports) &&
+		equality.Semantic.DeepEqual(desiredSvc.Spec.Selector, found.Spec.Selector) &&
+		equality.Semantic.DeepEqual(desiredSvc.Spec.Type, found.Spec.Type) &&
+		equality.Semantic.DeepEqual(desiredSvc.Spec.SessionAffinity, found.Spec.SessionAffinity)
```
Comment (cliveseldon) [inline operator/controllers/seldondeployment_controller.go:883]: There are several other fields that can be set in the spec by the user: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#servicespec-v1-core
Have you validated none of these is applicable? Some have defaults and are usually not set so I suppose one would need to be careful if as we are doing now we are not copying the spec back.

Comment (caozhuozi) [reply operator/controllers/seldondeployment_controller.go:883]: @cliveseldon, Thanks for you quick review. Do you mean in addition to the defaults set by Kubernetes, we also need to consider that the users may (intentionally or unintentionally) modify the services directly?

Comment (cliveseldon) [reply operator/controllers/seldondeployment_controller.go:883]: Actually, as a user has very little control from  SeldonDeployment on the created service we can ignore fields like `externalIPs` etc as a user can't control these from Seldon.

Comment (caozhuozi) [reply operator/controllers/seldondeployment_controller.go:883]: @cliveseldon. If we can ignore the impact of users, can I say that the change of svc fields are controlled by those defaults by Kubernetes and seldon operator? All the fields I selected to comapre/update are extracted from _createContainerServices_ and _ceratePredictorServices_ which are required by seldon operator side. I think keeping the defaults set by Kubernetes shouldn't bring side effects. Could you please give specific situations that we have to take other fields into account?

Comment (cliveseldon) [reply operator/controllers/seldondeployment_controller.go:883]: No I think that is fine if we have covered all the items in the Spec that we control

Comment (cliveseldon) [reply operator/controllers/seldondeployment_controller.go:883]: I think its ok as is

**Review by cliveseldon — APPROVED** [review]

**Pull request merged**